test/test_images/latencymako/sender_receiver.go

```diff
@@ -194,38 +126,112 @@ func main() {
 	printf("Estimated total messages: %v", estimatedNumberOfTotalMessages)
 
 	// Create all channels
-	sentCh = make(chan sentState, estimatedNumberOfMessagesInsideAChannel)
-	acceptedCh = make(chan acceptedState, estimatedNumberOfMessagesInsideAChannel)
-	failedCh = make(chan failedState, estimatedNumberOfMessagesInsideAChannel)
-	receivedCh = make(chan receivedState, estimatedNumberOfMessagesInsideAChannel)
+	executor.sentCh = make(chan sentState, estimatedNumberOfMessagesInsideAChannel)
+	executor.acceptedCh = make(chan acceptedState, estimatedNumberOfMessagesInsideAChannel)
+	executor.failedCh = make(chan failedState, estimatedNumberOfMessagesInsideAChannel)
+	executor.receivedCh = make(chan receivedState, estimatedNumberOfMessagesInsideAChannel)
 
 	// Small note: receivedCh depends on receive thpt and not send thpt but we
 	// don't care since this is a pessimistic estimate and receive thpt < send thpt
 
-	sentEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	acceptedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	failedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	receivedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
+	// Initialize records maps
+	executor.sentEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_SENT,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.acceptedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_ACCEPTED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.failedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_FAILED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.receivedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_RECEIVED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+
+	// --- Create Vegeta attacker
+
+	client := http.Client{Transport: requestInterceptor{
+		before: func(request *http.Request) {
+			id := request.Header.Get("Ce-Id")
+			executor.sentCh <- sentState{eventId: id, at: ptypes.TimestampNow()}
+		},
+		after: func(request *http.Request, response *http.Response, e error) {
+			id := request.Header.Get("Ce-Id")
+			t := ptypes.TimestampNow()
+			if e != nil || response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
+				executor.failedCh <- failedState{eventId: id, at: t}
+			} else {
+				executor.acceptedCh <- acceptedState{eventId: id, at: t}
+			}
+		},
+	}}
+
+	executor.attacker = vegeta.NewAttacker(
+		vegeta.Client(&client),
+		vegeta.Workers(workers),
+		vegeta.MaxWorkers(workers),
+	)
+
+	return executor
+}
+
+type requestInterceptor struct {
```

Comment:
Can be fixed in a later PR, but this should probably move to its own file, since it's generic and can be tested separately.